torch/csrc/jit/tensorexpr/cuda_codegen.cpp

```diff
@@ -749,7 +751,38 @@ Stmt* GPUMetaVarRewriter::mutate(const For* v) {
   return v->cloneWithNewBody(body);
 }
 
+std::vector<Stmt*> GPUMetaVarRewriter::handleDependences(const Block* v) {
+  analysis::MemDependencyChecker analyzer;
+  v->accept(&analyzer);
+
+  // Iterate over the statements in the block. If any statement is dependent on
+  // any previous statement in the block, then insert a SyncThreads before that
+  // statement. Note that the iteration below is quadratic in the number of
+  // statements in the block.
+  auto vStmts = v->stmts();
+  std::vector<Stmt*> stmts(vStmts.begin(), vStmts.end());
+  std::set<Stmt*> stmtsNeedingSync;
+  for (auto it1 = stmts.begin(); it1 != stmts.end(); ++it1) {
+    for (auto it2 = std::next(it1); it2 != stmts.end(); ++it2) {
+      auto hazard = getPotentialHazards(analyzer, *it1, *it2);
+      if (hazard != HazardKind::NoDependency) {
+        stmtsNeedingSync.insert(*it2);
+      }
+    }
+  }
+  std::vector<Stmt*> stmtsWithSync;
+  for (size_t i = 0; i < stmts.size(); ++i) {
+    if (stmtsNeedingSync.find(stmts[i]) != stmtsNeedingSync.end()) {
+      stmtsWithSync.push_back(new SyncThreads());
```

Comment:
I think you'll be inserting a lot of false positives here, e.g. 
```
for (int x = 0; x < 10; ++x) {
 A[x] = x;
 B[x] = A[x+1];
}
for (int y = 0; y < 10; ++y) {
 C[y] = A[y+1];
}
```

I think you'll insert two __syncthreads() - once between the store to A[x] and B[x], and once between the two loop stmts. Can you test this case?

Reply:
Good point. You were right that my initial implementation was indeed adding two `syncthreads`. 

I have updated the implementation to not add redundant sync threads now, like in this case. PTAL.